rfcs/0108-nixos-containers.md

```diff
@@ -0,0 +1,467 @@
+  * With this design, it won't be necessary to implement adjustments for advanced features
+    such as [MACVLAN interfaces](https://backreference.org/2014/03/20/some-notes-on-macvlanmacvtap/)
+    since administrators can directly use the upstream configuration format. The current module
+    supports `MACVLAN` interfaces for instance, but not `IPVLAN`.
+  * Another side effect is that existing knowledge about this configuration can be re-used.
+
+* Provide a reasonable upgrade path for existing installations. Even though this RFC suggests
+  deprecating the existing `nixos-container` subsystem, this measure is purely optional. However,
+  for this to happen, a smooth migration path must be provided.
+
+# Detailed design
+[design]: #detailed-design
+
+## Bootstrapping
+
+To be fully consistent with upstream `systemd`, the template unit
+[`systemd-nspawn@.service`](https://github.com/systemd/systemd/blob/v247/units/systemd-nspawn@.service.in) will be used.
+
+The approach how a container is bootstrapped won't change and will thus consist of the
+following steps (executed via a custom `ExecStartPre=`-script):
+
+* Create an empty directory in `/var/lib/machines` named like the container-name.
+* `systemd-nspawn` only expects `/etc/os-release`, `/etc/machine-id` and `/var` to exist
+  inside, however with no content.
+* To get a running NixOS inside, `/nix/store` is bind-mounted into it. As an `init` process,
+  the [stage-2 script](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/stage-2-init.sh)
+  is started which eventually `exec(2)`s into `systemd` and ensures that everything
+  is correctly set up.
+* The option `boot.isContainer = true;` will be automatically set for new containers as well.
+  This is necessary to
+  * avoid bogus `modprobe` calls since `nspawn` doesn't have its own kernel.
+  * as option to skip building a `stage-1` boot script when building a NixOS system for
+    the container.
+
+This init-script can be built by evaluating a NixOS config against `<nixpkgs/nixos/lib/eval-config.nix>`.
+
+Support for existing tarballs to be imported with `machinectl pull-tar` is explicitly out of
+scope in this RFC.
+
+## Network
+
+The following section provides an overview of how to configure networking for containers and
+how this will be implemented. A proposal how the API of the NixOS module could look like will
+be demonstrated in the [next chapter](#examples-and-interactions).
+
+### "public" networking
+
+This is the most trivial networking mode. It is taken if the `PrivateNetwork`-option of the
+`.nspawn`-unit is set to `no`. In this case, the container has full access to the host's network,
+otherwise the container will run in its own namespace.
+
+### Default Mode
+
+If nothing else is specified, the [default settings of `systemd-nspawn`](https://github.com/systemd/systemd/blob/v247/network/80-container-ve.network) will
+be used for networking. To briefly summarize, this means:
```

Comment:
I'm not sure if we should be too specific with the networkd config format internals here, and just describe how it'll look like from a birds-eye perspective.